### PR TITLE
Shared layers between tests.

### PR DESCRIPTION
**Pull request by swoogles**
Currently, when the normal `run` method in `ZIOSpec` is called, it finishes by calling `sys.exit`.
This behaves fine with ZIO 1's `AbstractRunnableSpec`, but the approach I'm taking with 2.0's `ZIOSpecAbstract` kills SBT and `ZTestFrameworkSpec`.

### CODE REVIEW

File: core-tests/shared/src/test/scala/zio/sharedLayers/SharedLayer.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+import zio.{Has, ZIO, ZLayer}
+
+import java.util.concurrent.atomic.AtomicInteger
+
+object SharedLayer {
+  private val counter = new AtomicInteger(0)
+
+  val test: ZLayer[Any, Nothing, Has[Unit]] = {
+    ZLayer.fromZIO(ZIO.debug("Constructing a super expensive layer") *> ZIO.succeed(updateCounter()))
```
Comment (adamgfraser) [inline core-tests/shared/src/test/scala/zio/sharedLayers/SharedLayer.scala:11]: We should probably remove this debug statement.

File: test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala [hunk 2]
```diff
@@ -34,12 +36,13 @@ object ZTestFrameworkSpec {
     test("should correctly display colorized output for multi-line strings")(testColored()),
     test("should test only selected test")(testTestSelection()),
     test("should return summary when done")(testSummary()),
+    test("should use a shared layer without re-initializing it")(testSharedLayers()),
```
Comment (adamgfraser) [inline test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala:39]: Are we still using this or are we now just testing by having those two tests in the normal ZIO Test suite?

File: test/shared/src/main/scala/zio/test/DefaultRunnableSpec.scala [hunk 3]
```diff
@@ -37,8 +37,10 @@ abstract class DefaultRunnableSpec extends RunnableSpec[TestEnvironment, Any] {
    */
   private[zio] override def runSpec(
     spec: ZSpec[Environment, Failure]
-  )(implicit trace: ZTraceElement): URIO[Has[TestLogger] with Has[Clock], ExecutedSpec[Failure]] =
+  )(implicit trace: ZTraceElement): URIO[Has[TestLogger] with Has[Clock], ExecutedSpec[Failure]] = {
+    println("DefaultRunnableSpec.runSpec")
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/DefaultRunnableSpec.scala:41]: Stray debug statement?

File: test/shared/src/main/scala/zio/test/NewOrLegacySpec.scala [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package zio.test
+
+sealed trait NewOrLegacySpec
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/NewOrLegacySpec.scala:3]: Does this need to be in ZIO Test or can it be moved to ZIO Test SBT?

File: test/shared/src/main/scala/zio/test/RunnableSpec.scala [hunk 5]
```diff
@@ -29,6 +29,7 @@ abstract class RunnableSpec[R, E] extends AbstractRunnableSpec {
   private def run(spec: ZSpec[Environment, Failure], testArgs: TestArgs)(implicit
     trace: ZTraceElement
   ): URIO[Has[TestLogger] with Has[Clock], Int] = {
+    println("RunnableSpec.run")
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/RunnableSpec.scala:32]: Debug statement.

File: test/shared/src/main/scala/zio/test/RunnableSpec.scala [hunk 6]
```diff
@@ -60,9 +61,11 @@ abstract class RunnableSpec[R, E] extends AbstractRunnableSpec {
     }
   }
 
-  private def doExit(exitCode: Int): Unit =
+  private def doExit(exitCode: Int): Unit = {
+    println("About to exit legacy")
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/RunnableSpec.scala:65]: Debug statement.

File: test/shared/src/main/scala/zio/test/ZIOSpec.scala [hunk 7]
```diff
@@ -19,7 +19,10 @@ package zio.test
 import zio._
 import zio.stacktracer.TracingImplicits.disableAutoTrace
 
-abstract class ZIOSpec[R <: Has[_]: Tag] extends ZIOApp { self =>
+import org.portablescala.reflect.annotation.EnableReflectiveInstantiation
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/ZIOSpec.scala:22]: Do we need to do this here? I think it is sufficient to do it at the `ZIOSpecAbstract` level.

File: test/shared/src/main/scala/zio/test/ZIOSpecAbstract.scala [hunk 8]
```diff
@@ -56,10 +57,16 @@ abstract class ZIOSpecAbstract extends ZIOApp { self =>
...
-      _        <- doExit(exitCode)
+      args         <- ZIO.service[ZIOAppArgs]
+      testArgs      = TestArgs.parse(args.getArgs.toArray)
+      executedSpec <- runSpec(spec, testArgs)
+      hasFailures = executedSpec.exists {
+                      case ExecutedSpec.TestCase(test, _) => test.isLeft
+                      case _                              => false
+                    }
+      exitCode = if (hasFailures) 1 else 0
+      _       <- ZIO.debug("About to exit")
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/ZIOSpecAbstract.scala:68]: Debug statement.

File: test/shared/src/main/scala/zio/test/ZIOSpecAbstract.scala [hunk 9]
```diff
@@ -20,6 +20,7 @@ import org.portablescala.reflect.annotation.EnableReflectiveInstantiation
 import zio._
 import zio.internal.stacktracer.Tracer
 import zio.stacktracer.TracingImplicits.disableAutoTrace
+import zio.test.ExecutedSpec.LabeledCase
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/ZIOSpecAbstract.scala:23]: Stray import?

**Review by adamgfraser — CHANGES_REQUESTED** [review]
A couple of clean up comments but looks really good!

File: core-tests/shared/src/test/scala/zio/sharedLayers/SharedLayer.scala (hunk 1 above)
Comment (swoogles) [reply core-tests/shared/src/test/scala/zio/sharedLayers/SharedLayer.scala:11]: Yep, removed

File: test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala (hunk 2 above)
Comment (swoogles) [reply test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala:39]: Removed the non-obviously-linked `RequriesLayer*Spec` that performed the same check.
We'll strictly rely on this test to confirm basic layer sharing now.

File: test/shared/src/main/scala/zio/test/DefaultRunnableSpec.scala (hunk 3 above)
Comment (swoogles) [reply test/shared/src/main/scala/zio/test/DefaultRunnableSpec.scala:41]: Removed!

File: test/shared/src/main/scala/zio/test/RunnableSpec.scala (hunk 5 above)
Comment (swoogles) [reply test/shared/src/main/scala/zio/test/RunnableSpec.scala:32]: Removed!

File: test/shared/src/main/scala/zio/test/RunnableSpec.scala (hunk 6 above)
Comment (swoogles) [reply test/shared/src/main/scala/zio/test/RunnableSpec.scala:65]: Removed!

File: test/shared/src/main/scala/zio/test/ZIOSpecAbstract.scala (hunk 8 above)
Comment (swoogles) [reply test/shared/src/main/scala/zio/test/ZIOSpecAbstract.scala:68]: Removed

File: test/shared/src/main/scala/zio/test/ZIOSpecAbstract.scala (hunk 9 above)
Comment (swoogles) [reply test/shared/src/main/scala/zio/test/ZIOSpecAbstract.scala:23]: Yep, removed.

File: test/shared/src/main/scala/zio/test/ZIOSpec.scala (hunk 7 above)
Comment (swoogles) [reply test/shared/src/main/scala/zio/test/ZIOSpec.scala:22]: Great catch, works fine when applied only to `ZIOSpecAbstract`

File: test/shared/src/main/scala/zio/test/NewOrLegacySpec.scala (hunk 4 above)
Comment (swoogles) [reply test/shared/src/main/scala/zio/test/NewOrLegacySpec.scala:3]: Another good catch- moved!

**Comment by swoogles** [comment]
I can't figure what's going wrong with Scala 3 here.
```
 Exception in thread "zio-fiber-1635739907" java.lang.Error: Defect in zio.Has: HashSet({Has[=Clock] & Has[=Console] & Has[=Random] & Has[=System]}) statically known to be contained within the environment are missing
 ```
 The offending block is:
 ```
   protected def run(
    eventHandler: EventHandler,
    spec: ZIOSpecAbstract,
    loggers: Array[Logger]
  )(implicit trace: ZTraceElement): ZIO[Any, Throwable, Unit] =
    (for {
      spec <- spec
                .runSpec(FilteredSpec(spec.spec, args), args)
                .provideLayer(
                  spec.layer.mapError(e => new Error(e.toString)) >+> ZLayer.succeed(
                    ZIOAppArgs(Chunk.empty)
                  ) >+> zio.ZEnv.live >+> sbtTestLayer(loggers) >+> TestEnvironment.live
                )
      events = ZTestEvent.from(spec, taskDef.fullyQualifiedName(), taskDef.fingerprint())
      _     <- ZIO.foreach(events)(e => ZIO.attempt(eventHandler.handle(e)))
    } yield ()).provideLayer(ZLayer.succeed(ZIOAppArgs(Chunk.empty)) >+> zio.ZEnv.live)
    ```
    at `runSpec.provideLayer`.
    
    Spent the last few hours on this one failure, to no avail.
    Is there some reduced capability of Scala 3's `Has` that I should be aware of?

**Review by adamgfraser — APPROVED** [review]
Fantastic work!

**Pull request merged**